File: catowserAndroid/browser-content/src/main/java/org/cotton/browser/content/service/tab/TabSelectionStrategy.kt
package org.cotton.browser.content.service.tab

/**
 * A Tab selection strategy (Compositor) defines the algorithms of tab selection in specific cases:
 * - when tab was removed and need to select another
 *
 * */
interface TabSelectionStrategy {
    /**
     * Computes the new index for the currently selected tab when that tab is removed
     *
     * @param context An index selection context, usually the tabs list manager
     * @param removedIndex The index which was removed
     * @return computed new index of the tab which needs to be selected
     * */
    suspend fun autoSelectedIndexAfterTabRemove(context: IndexSelectionContext, removedIndex: Int): Int
    /**
     * Tells if the app needs to select the new tab after it is added
     * */
    val makeTabActiveAfterAdding: Boolean
}

File: catowseriOS/Domain/Sources/CottonTabs/TabSelectionStrategy.swift
//
//  TabSelectionStrategy.swift
//  CottonDataServices
//
//  Created by Andrei Ermoshin on 05/03/2019.
//  Copyright © 2019 Cotton/Catowser Andrei Ermoshin. All rights reserved.
//

import AutoMockable

public protocol IndexSelectionContext: AutoMockable {
    var collectionLastIndex: Int { get async }
    var currentlySelectedIndex: Int { get async }
}

/// Snapshot-based `IndexSelectionContext` for selection strategy calls outside `TabsDataService`.
public struct TabsIndexSelectionSnapshot: IndexSelectionContext, Sendable {
    private let lastIndex: Int
    private let selectedIndex: Int

    public init(collectionLastIndex: Int, currentlySelectedIndex: Int) {
        self.lastIndex = collectionLastIndex
        self.selectedIndex = currentlySelectedIndex
    }

    public var collectionLastIndex: Int {
        get async { lastIndex }
    }

    public var currentlySelectedIndex: Int {
        get async { selectedIndex }
    }
}

/// CoreBrowser.Tab selection protocol can be sendable, because implementation
/// only holds a constant which can't be mutated, so that, no any mutable state for now.
public protocol TabSelectionStrategy: AutoMockable, Sendable {
    /// A CoreBrowser.Tab selection strategy (Compositor) defines the algorithms of tab selection in specific cases
    /// 1) when tab was removed and need to select another
    /// 2) tbd
    ///
    /// - Parameter context: An additional info required to do the index calculations
    /// - Parameter removedIndex: Index which was removed
    /// - Returns a new selected index or nil if removed tab wasn't selected
    func autoSelectedIndexAfterTabRemove(
        context: IndexSelectionContext,
        removedIndex: Int
    ) async -> Int?
    /// Shows if we need to make the tab active/selected after creating it
    var makeTabActiveAfterAdding: Bool { get }
}
